### DISCUSSION TITLE
Promise methods `then` and `map` don't handle Exceptions

### DISCUSSION THREAD

**Discussion by Qudratillo** [activej/activej#167] [Ideas]
I am very excited to use ActiveJ and I see it is incredible project

I got a problem that `Promise::then` and `Promise::map` simply ignores exceptions thrown inside handlers (fn) given to them and this causes promises to stop to call next callbacks.

Here inside `CallbackList`
```
public void accept(T result, @Nullable Throwable e) {
	for (int i = 0; i < index; i++) {
		callbacks[i].accept(result, e);
	}
}
```
I think there should be a catch to exceptions otherwise `for` loop just stops when any of callbacks throws an exception

**Comment by eduard-vasinskyi** [comment]
Hi, @Qudratillo

`Promise::then` and `Promis::map` do indeed ignore exceptions. To handle exceptions there are ‘Ex’-methods `mapEx` and `thenEx` that take up `BiFunction<T, Throwable, U>` and `BiFunction<T, Throwable, Promise<U>>` respectively. 

You would need to inspect whether a Throwable is `null` or not. If a Throwable  is `null`, that means the promise has completed successfully, otherwise there was an error. Here is a simple snippet:

```java
someAsyncTask()
     .thenEx((result, e) -> {
        if (e == null) {
           return someAsyncTask2();
        } else {
           return Promise.ofException(new MyException(e));
        }
     });
 ```
 
If you do not handle the exception properly, the next `map()`, `then()`, `whenResult()` methods in the promise chain would never be called. Ideally, you would need to have some ‘Ex’-method handler at the end of your promise chain in order to not lose an exception.

As for the loop in `CallbackList`, there should not be a try-catch block because callbacks are not expected to throw exceptions. Callbacks should rather return an exceptionally completed promise using `Promise.ofException(Throwable)` (as in the above example). If inside a callback you use an API that may throw an exception, you would need to do all the try-catching inside your callback lambda. 

If an unchecked exception is thrown from within a callback, it indeed would break a promise chain execution. The current eventloop would intercept such unchecked exception and would pass it to the `Eventloop#fatalErrorhandler` (which would ignore all exceptions by default). You can change the handling of fatal errors by providing your own handler to the eventloop using `Eventloop#withFatalErrorHandler`. There are also some common handlers in the `FatalErrorHandlers` class.
You can read more about promises here, there are links to examples as well.

**Comment by Qudratillo** [comment]
Hi @eduard-vasinskyi 
Thank you for answering

I made some workarounds using `then` functions
```
fun <T, O> Promise<T>.thenCatching(fn: Function<T, Promise<O>>) = then { r ->
    try {
        fn.apply(r)
    } catch (e: Throwable) {
        Promise.ofException(e)
    }
}

fun <T, O> Promise<T>.mapCatching(fn: Function<T, O>) = then { r ->
    try {
        Promise.of(fn.apply(r))
    } catch (e: Throwable) {
        Promise.ofException(e)
    }
}

fun <T> Promise<T>.whenResultCatching(fn: Consumer<T>) = then { r ->
    try {
        fn.accept(r)
        this
    } catch (e: Throwable) {
        Promise.ofException(e)
    }
}
```

However, I feel like there is an extra overhead since these functions create new Promises.

**Comment by eduard-vasinskyi** [comment]
@Qudratillo
We have redesigned exception handling in promises based on your suggestions. The changes are currently part of the ActiveJ v5.0-beta2 release, currently on maven.

Previous ‘Ex’-methods `mapEx()`, `thenEx()` have been renamed to `map()` and `then()`. Now all of these methods (`map()`, `then()` `whenComplete()`, `whenResult()`, `whenException()`) take ‘throwing’ functions. Whenever a code in such function throws a checked exception, the whole promise is completed exceptionally. Any unchecked exception would be thrown as-is, without completing the promise.

**Comment by Qudratillo** [comment]
@eduard-vasinskyi 
Thank you

It is a good solution, but I think unresponsive promises may still trouble developers. Maybe you should change it like when `Error` or `RuntimeException` is thrown promises complete exceptionally and `Eventloop` still records fatal error.
Promises are core part of reactive servers and unresponsive promises leads to either unresponsiveness or timeout errors by these servers. Instead the servers may response immediately with `Internal Server Error`
If promises complete exceptionally when any error is thrown, testing time also decreases (if there is an error a test fails immediately otherwise it needs to wait until it runs out of time).

**Comment by eduard-vasinskyi** [comment]
Hi, @Qudratillo
Thank you for the suggestion. 

We have decided to make handling fatal errors more flexible and configurable. We have abstracted out a [FatalErrorHandler](https://github.com/activej/activej/blob/master/util-common/src/main/java/io/activej/common/exception/FatalErrorHandler.java) class that operated solely on `Eventloop` level to be a generic solution for handling fatal errors that occur in promises or any other part of the code. 

Take a look at [FatalErrorHandlers](https://github.com/activej/activej/blob/master/util-common/src/main/java/io/activej/common/exception/FatalErrorHandlers.java) class. It contains a thread local of fatal error handlers as well as a global fatal error handler which is used as a default handler if none thread handler has been set. Whenever an unchecked exception is thrown in promises, it is handled by either a thread-level handler or a default global error handler.

You may set thread-level handler manually by calling `FatalErrorHandlers#setThreadFatalErrorHandler`. Alternatively, you can set thread handler when creating Eventloop (`Eventloop#withThreadFatalErrorHandler`). 

Eventloop actually allows 2 fatal error handlers to be set. One on a thread level and another final handler for handling any errors rethrown by a thread level handler.

The default behavior is the same as it was in ActiveJ v4, that is any unchecked exception is rethrown on a thread level (in promises, etc.). Such unchecked exceptions are then caught, logged, and recorded on the eventloop level.

You can change this behavior to the one you suggested by calling `eventloop.withThreadFatalErrorHandler(FatalErrorHandler.logging())`. The next time an unchecked exception is thrown inside the promise `then()/map()/whenXXX()` method, it will be logged and a promise will be completed exceptionally with an unchecked exception.

You can also take a look at [FatalErrorHandler](https://github.com/activej/activej/blob/master/util-common/src/main/java/io/activej/common/exception/FatalErrorHandler.java) interface as it is now the basis for handling unchecked exceptions. It allows chaining handlers together by calling `FatalErrorHandler#andThen` method. There are also multiple predefined handlers (such as `logging()`) that you can use.

This change is a part of the latest ActiveJ v5.0-rc3 release candidate. This version has already been tested in our production applications, so a final ActiveJ v5.0 is planned to be released in the next few days.

**Comment by Qudratillo** [comment]
@eduard-vasinskyi Thank you
